compiler/codegen/src/string_parser.rs

```diff
@@ -0,0 +1,287 @@
+//! A stripped-down version of ruff's string literal parser, modified to
```

Comment:
Would ruff parser also need to be patched?

Reply:
Yeah, though I'm not sure whether they'd necessarily feel a need to. I think what might strike a good balance is if there's a flag that can be set like `contains_surrogates`, so we can know whether or not we need to reparse.

Reply:
Yeah, probably. There's an open issue but it isn't something that has come up often. https://github.com/astral-sh/ruff/issues/13666